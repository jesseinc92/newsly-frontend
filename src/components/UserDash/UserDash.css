.UserDash {
  margin-top: 7rem;
}

.UserDash h1,
.UserDash h2 {
  font-family: 'Fjalla One', sans-serif;
}

.UserDash h1 {
  font-size: 2.5rem;
}

.UserDash-metrics-box {
  margin-top: 4rem;
}

.UserDash-metrics-box select {
  font-size: 1.25rem;
  font-family: 'taviraj', serif;
  padding-left: 0.5rem;
  margin-top: 1rem;
  margin-bottom: 2rem;
}

.metrics-graph {
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 1rem;

  width: 85%;
  margin: auto;
  margin-top: 6rem;
}

.metrics-bar {
  display: none;
}

.metrics-bundle {
  display: flex;
  flex-direction: column;
}

.UserDash-metrics-cat {
  border: 2px solid;
  border-radius: 1.25rem;
  padding: 0.25rem 1rem;
  grid-column: 1;
  /* background-color: unset; */
}

.metrics-interpretation {
  text-align: left;
  width: 75%;
  margin: auto;
}

.UserDash-delete-box {
  display: flex;
  flex-direction: column;
  width: 75%;
  padding: 2rem;
  margin: auto;
  margin-top: 3rem;
  margin-bottom: 2rem;

  border: 1px solid #495057;
  border-radius: 1rem;
}

.UserDash-delete-box span {
  color: #495057;
  text-align: left;
}


/* Tablet sizing */
@media screen and (min-width: 768px) {

  .UserDash h1 {
    font-size: 3.5rem;
  }

  .UserDash h2 {
    font-size: 2.5rem;
  }

  .metrics-graph {
    width: 75%;
    gap: 2rem;
  }

}

/* Laptop/desktop sizing */
@media screen and (min-width: 1024px) {

  .UserDash h1 {
    font-size: clamp(48px, 4.69vw, 64px);
    text-align: left;
    padding-left: 5rem;
  }

  .UserDash h2 {
    font-size: clamp(40px, 3.9vw, 48px);
  }

  .metrics-bundle {
    font-size: 1rem;
    flex-direction: column-reverse;
  }

  .metrics-graph {
    grid-template-columns: repeat(8, 1fr);
    gap: 1rem;
  }

  .metrics-bar {
    display: inline-block;
    width: 1.5vw;
    max-height: 20rem;
    margin: 0 auto;
    margin-top: auto;
    background-color: #868e96;
    border-radius: 1rem;
  }

  .metrics-interpretation {
    width: 40%;
  }

  .UserDash-delete-box {
    width: 50%;
  }

}