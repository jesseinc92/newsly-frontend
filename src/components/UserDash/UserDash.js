import React, { useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import UserContext from '../../UserContext';
import useFormChange from '../../hooks/useFormChange';
import interpretMetrics from '../../helpers/interpretMetrics';
import NewslyAPI from '../../api';
import './UserDash.css';
import Button from '../Button/Button';

const UserDash = () => {
  const redirect = useNavigate();
  const { user, setUser } = useContext(UserContext);
  const [edit, setEdit] = useState(false);
  const [goal, setGoal] = useState(false);

  const showEditForm = () => {
    setEdit(edit => !edit);
  }

  const showGoalForm = () => {
    setGoal(edit => !edit);
  }

  const [formData, handleChange] = useFormChange({
      username: user.username,
      firstName: user.firstName,
      lastName: user.lastName
  });

  const [goalData, handleGoalChange] = useFormChange({ goal: user.metrics.goal })

  const editUser = async (e) => {
    e.preventDefault();
    try  {
      const updatedUser = await NewslyAPI.updateUser(formData);
      
      localStorage.setItem('user', JSON.stringify(updatedUser));
      setUser(updatedUser);
      setEdit(false);
    } catch(err) {
      alert(err);
    }
  }

  const updateGoal = async (e) => {
    e.preventDefault();
    console.log(goalData)
    try {
      const updatedGoal = await NewslyAPI.updateGoal(user.username, goalData);
      const updatedUser = await NewslyAPI.getUser(user.username)

      setUser(updatedUser);
      localStorage.setItem('user', JSON.stringify(updatedUser));
      setGoal(false);
    } catch(err) {
      alert(err);
    }
  }

  const deleteUser = async () => {
    try {
      const deleteMessage = await NewslyAPI.deleteUser(user.username);
      alert(deleteMessage);

      // Logout user after they are deleted from database
      localStorage.removeItem('user');
      setUser(null);
      redirect('/');
    } catch(err) {
      alert(err);
    }
  }

  return (
    <section className='UserDash'>
      <h2>Hello, { user.firstName }</h2>

      <div className='UserDash-prof-details'>
        <h4>User Details</h4>
        <Button text='Edit' handler={showEditForm} />

        { edit ? 
          <form onSubmit={editUser}>
            <div>
              <label htmlFor='firstName'><strong>First Name: </strong></label>
              <input 
                type='text' 
                id='firstName' 
                name='firstName' 
                value={formData.firstName} 
                onChange={handleChange} 
                autoFocus 
              />
            </div>
            <div>
              <label htmlFor='lastName'><strong>Last Name: </strong></label>
              <input 
                type='text' 
                id='lastName' 
                name='lastName' 
                value={formData.lastName} 
                onChange={handleChange} 
              />
            </div>
            <Button text='Save' />
          </form>
          :
          <>
            <p><strong>Username:</strong> { user.username }</p>
            <p><strong>First Name:</strong> { user.firstName }</p>
            <p><strong>Last Name:</strong> { user.lastName }</p>
          </>
        }
      </div>

      
      <div className='UserDash-metrics-box'>
        <h3>Your News by the Numbers</h3>
        <Button text='Update' handler={showGoalForm} />

        { goal ?
          <form onSubmit={updateGoal}>
            <div>
              <label htmlFor='goals'><strong>I want my goal to be: </strong></label>
              <select name='goal' id='goals' value={goalData.goal} onChange={handleGoalChange}>
                <option value='all'>All</option>
                <option value='usNews'>US News</option>
                <option value='worldNews'>World News</option>
                <option value='business'>Business</option>
                <option value='opinion'>Opinion</option>
                <option value='sport'>Sport</option>
                <option value='culture'>Culture</option>
                <option value='science'>Science</option>
                <option value='lifestyle'>Lifestyle</option>
              </select>
            </div>
            <Button text='Save'/>
          </form>
          :
          <>
            <div className='metrics-graph'>
              <p>You want to { user.metrics.goal ?? 'undecided' }.</p>
              <p><span className='UserDash-metrics-cat usNews'>US News</span> { user.metrics.usNews }</p>
              <p><span className='UserDash-metrics-cat worldNews'>World News</span> { user.metrics.worldNews }</p>
              <p><span className='UserDash-metrics-cat business'>Business</span> { user.metrics.business }</p>
              <p><span className='UserDash-metrics-cat opinion'>Opinion</span> { user.metrics.opinion }</p>
              <p><span className='UserDash-metrics-cat sport'>Sport</span> { user.metrics.sport }</p>
              <p><span className='UserDash-metrics-cat culture'>Culture</span> { user.metrics.culture }</p>
              <p><span className='UserDash-metrics-cat science'>Science</span> { user.metrics.science }</p>
              <p><span className='UserDash-metrics-cat lifestyle'>Lifestyle</span> { user.metrics.lifestyle }</p>
            </div>
            <div className='metrics-interpretation'>
              <p>{ interpretMetrics(user.metrics) }</p>
            </div>
          </>
        }
      </div>

      <div className='UserDash-delete-box'>
        <Button text='Delete Account' handler={deleteUser} />
        <span>
          Doing this will remove all of your bookmarked articles and will erase 
          all of your saved goal progress. Be sure this is what you want to do, 
          there's no going back.
        </span>
      </div>
    </section>
  )
}

export default UserDash;