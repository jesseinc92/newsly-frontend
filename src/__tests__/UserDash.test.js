import React from "react";
import { MemoryRouter } from "react-router-dom";
import { screen } from '@testing-library/react'
import renderWithContext from '../helpers/testContext';
import UserDash from '../components/UserDash/UserDash';

it('renders the component', function() {
  renderWithContext(<MemoryRouter><UserDash /></MemoryRouter>)
});

it('matches the snapshot', function() {
  const { asFragment } = renderWithContext(<MemoryRouter><UserDash /></MemoryRouter>);
  expect(asFragment()).toMatchSnapshot();
});

it('displays the user name', function() {
  renderWithContext(<MemoryRouter><UserDash /></MemoryRouter>);
  const greeting = screen.getByText('Hello, test');
  expect(greeting).toBeInTheDocument();
})