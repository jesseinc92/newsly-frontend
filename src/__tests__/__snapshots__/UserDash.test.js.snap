// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <section
    class="UserDash"
  >
    <h1>
      Hello, test
    </h1>
    <div
      class="UserDash-metrics-box"
    >
      <h2>
        Your News by the Numbers
      </h2>
      <p
        class="metrics-graph-goal"
      >
        You want to view more Goal.
      </p>
      <div
        class="metrics-graph"
      >
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat usNews"
          >
            US
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat worldNews"
          >
            World
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat business"
          >
            Business
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat opinion"
          >
            Opinion
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat sport"
          >
            Sport
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat culture"
          >
            Culture
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat science"
          >
            Science
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
        <div
          class="metrics-bundle"
        >
          <span
            class="UserDash-metrics-cat lifestyle"
          >
            Lifestyle
          </span>
          <span />
          <span
            class="metrics-bar"
          />
        </div>
      </div>
      <div
        class="metrics-interpretation"
      >
        <p>
          You need to read more articles before we can evaluate your goal progress.
        </p>
      </div>
      <button
        class="Button"
      >
        <a
          href="/"
        >
          Update Goal
        </a>
      </button>
      <button
        class="Button"
      >
        <a
          href="/"
        >
          Edit User Details
        </a>
      </button>
    </div>
    <div
      class="UserDash-prof-details"
    />
    <div
      class="UserDash-delete-box"
    >
      <button
        class="Button"
      >
        <a
          href="/"
        >
          Delete Account
        </a>
      </button>
      <span>
        Doing this will remove all of your bookmarks and will erase all of your metrics. This can't be undone!
      </span>
    </div>
  </section>
</DocumentFragment>
`;
